File: 2020/2020W29/preppindataw29.R
library(dplyr)
library(readxl)
library(tidyr)
library(stringr)

xlsx <- "F:/Data/Attendee List.xlsx"

finalA <- read_excel(xlsx, sheet = 'Attendee List') %>%
  mutate('Currency' = case_when(`Country`=='United States' ~ 'USD',
                                `Country`=='Canada' ~ 'CAD',
                                `Country`=='Mexico' ~ 'MXN',
                                `Country`=='United Kingdom' ~ 'GBP',
                                TRUE ~ 'EUR'),
         'Lower Company' = tolower(str_extract(`Email`,'(?<=@)([^\\.]*)(?=\\.)'))) %>%
  merge(read_excel(xlsx, sheet = 'Account Manager') %>%
          transform(.,'Order'=as.numeric(factor(`Account Manager`))) %>%
          rename('Company Name' = !!names(.[1]), 'Account Manager' = !!names(.[2])) %>%
          mutate('Lower Company' = tolower(`Company Name`)), by='Lower Company') %>%
  merge(read_excel(xlsx, sheet = 'Exchange Rates') %>%
          mutate('Currency' = str_extract(`Currency`,'^\\w{3}')) %>%
          select(-'GBP') %>%
          rbind(., data.frame('Currency'='GBP', 'Rate'=1)), by='Currency') %>%
  mutate('Ticket Price Local' = `Ticket Price (¢G)`*`Rate`) %>%
  select(c('First Name', 'Last Name', 'Email', 'Country', 'Refund Type',
           'Currency', 'Company Name', 'Account Manager', 'Order','Ticket Price Local'))
  
finalB <- finalA %>%
  group_by(`Country`,`Currency`) %>%
  summarize('Money Gain/Loss' = sum(`Ticket Price Local`*if_else(`Refund Type`=='Full Refund',-1,1)))
  
View(finalA)
View(finalB)

File: 2020/2020W29/PreppindataW29.py
import pandas as pd
import re

xlsx = pd.ExcelFile("F:/Data/Attendee List.xlsx")

attd = pd.read_excel(xlsx,'Attendee List')   
attd['Currency'] = attd['Country'].apply(lambda x: 'USD' if x=='United States' else
                                         'CAD' if x=='Canada' else
                                         'MXN' if x=='Mexico' else
                                         'GBP' if x=='United Kingdom' else 'EUR') 
attd['Lower Company'] = attd['Email'].apply(lambda x: re.search('(?<=@)([^\.]*)(?=\.)',x).group(1).lower())

mngr = pd.read_excel(xlsx,'Account Manager').rename(columns={'Company List':'Company Name'})
mngr_order = {mgr:idx+1 for idx, mgr in enumerate(list(set(mngr['Account Manager'].to_list())))}
mngr['Order'] = mngr['Account Manager'].apply(lambda x: mngr_order[x])
mngr['Lower Company'] = mngr['Company Name'].str.lower()  


rate = pd.read_excel(xlsx,'Exchange Rates').drop('GBP', axis=1)
rate['Currency'] = rate['Currency'].apply(lambda x: x[:3])
rate = pd.concat([rate, pd.DataFrame({'Currency': ['GBP'], 'Rate': [1]})])

finalA = attd.merge(mngr, on='Lower Company').drop('Lower Company',axis=1)
finalA = finalA.merge(rate, on='Currency')
finalA['Ticket Price Local'] = finalA['Ticket Price (£)']*finalA['Rate']
finalA = finalA.drop(['Ticket Price (£)','Rate'], axis=1)

finalB = finalA.copy()
finalB['Money Gain/Loss'] = finalB.apply(lambda x: (-1 if x['Refund Type']=='Full Refund' else 1)*x['Ticket Price Local'], axis=1)
finalB = finalB.groupby(['Country','Currency'], as_index=False).agg({'Money Gain/Loss':'sum'})
